### DISCUSSION TITLE
Mesa mixes up agent objects and runs another agent's `step` method

### DISCUSSION THREAD

**Discussion by jingxuanlim** [mesa/mesa#1291] [Q&A] [answered]
Hi. I ran into an issue that I feel like is a bug, because it works fine for multiple steps then stops working after.

Assume that I have two agent classes, `Predator` and `Prey` and they have different attributes and `step` methods. At this point, the two classes don't interact -- they just move. I found that they move fine for a variable number of steps, and then runs into the error (find output below, which ran for 10 steps but other times it would run for 6 or 1 etc). 

The traceback states that `Predator` is now trying to run `Prey`'s `step` method which relies on an attribute that only the `Prey` has (`velocity` in this case).

It therefore seems like mesa mixed up the two agent objects' methods, leading to an error. Has anyone experienced this?

It's possible that there's a problem with my code, which I am not able to find. Any hints welcomed!

```
WARNING:tornado.access:404 GET /favicon.ico (127.0.0.1) 3.50ms
{"type":"get_step","step":1}
{"type":"get_step","step":2}
{"type":"get_step","step":3}
{"type":"get_step","step":4}
{"type":"get_step","step":5}
{"type":"get_step","step":6}
{"type":"get_step","step":7}
{"type":"get_step","step":8}
{"type":"get_step","step":9}
{"type":"get_step","step":10}
ERROR:tornado.application:Uncaught exception GET /ws (127.0.0.1)
HTTPServerRequest(protocol='http', host='127.0.0.1:8521', method='GET', uri='/ws', version='HTTP/1.1', remote_ip='127.0.0.1')
Traceback (most recent call last):
.
.
.
AttributeError: 'Predator' object has no attribute 'velocity'
```

**Comment by rht** [comment, marked as answer]
If you have a GitHub repo I can checkout, it will be much easier to debug.
Edit: answer can be found in the thread below.

**Reply by jingxuanlim** [reply]
That would be awesome!

Repository is currently private but I've added you as a collaborator. Let me know if you didn't receive the invite.

The project is in the `schooling` subdir.

Thanks alot!!

**Reply by rht** [reply]
I found the cause. If the model is private, I can tell you at our [Matrix](https://matrix.to/#/#project-mesa:matrix.org).

**Reply by jingxuanlim** [reply]
Ah, no problem. I barely got started on it. Discussing in this thread is fine. Appreciate the thought though!

**Reply by rht** [reply]
The reason is because the `Prey` inherits from `Boid`. `Boid` has this method
```python
    def match_heading(self, neighbors):
        """
        Return a vector of the neighbors' average heading.
        """
        match_vector = np.zeros(2)
        if neighbors:
            for neighbor in neighbors:
                match_vector += neighbor.velocity
            match_vector /= len(neighbors)
        return match_vector
```
i.e. it doesn't check to make sure that the neighbor is a `Prey`. You have to override the `match_heading` method in the `Prey` class declaration, so that it checks whether the neighbor has `velocity` as an attribute.

**Reply by rht** [reply]
Actually, I think it's better to check by the class of the neighbors `isinstance(neighbor, Prey)`, rather than checking whether a neighbor has a `velocity` attribute.

**Reply by jingxuanlim** [reply]
Ahhh, silly me...! Of course I was wrong to assume that the agents are not interacting with each other! Works as expected now!

Also amazing that you were able to debug so quickly, despite having to look at someone else's code! Very impressed! Thanks!!

**Comment by jingxuanlim** [comment]
@rht One more thing I thought I ask here to get a quick answer. I hope you don't mind.

How do I get a video recording of the model running? Besides doing a screen recording, is there a way to do it from mesa?

**Reply by rht** [reply]
Hmm, never had to do that. I assume you want it like https://asciinema.org/ does for terminal sessions, where the recording output is semantic and much smaller than a video. Not sure this is the right path, but you might try extracting the output of the Chrome profiler (on Chrome/Chromium/Brave). You press ctrl-shift-I, and then go to the performance tab, and then you record. You should be able to see the rendered page being replayable.

**Reply by rht** [reply]
Agents.jl surely does have a nice feature having easily recordable animation: see https://juliadynamics.github.io/Agents.jl/v4.1/examples/sugarscape/. We might be able to do this using Matplotlib. But not sure how to do this given with custom agent portrayals that are not just circles.
